File: src/main/java/com/indiepost/mapper/ContributorMapper.java
package com.indiepost.mapper;

import com.indiepost.dto.ContributorDto;
import com.indiepost.model.Contributor;
import org.springframework.beans.BeanUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class ContributorMapper {
    public static ContributorDto contributorToContributorDto(Contributor contributor) {
        ContributorDto dto = new ContributorDto();
        BeanUtils.copyProperties(contributor, dto);
        return dto;
    }

    public static List<ContributorDto> contributorsToContributorDtos(List<Contributor> contributors) {
        if (contributors == null) {
            return new ArrayList<>();
        }
        return contributors.stream()
                .map(contributor -> contributorToContributorDto(contributor))
                .collect(Collectors.toList());
    }

    public static Contributor contributorDtoToContributor(ContributorDto dto) {
        Contributor contributor = new Contributor();
        BeanUtils.copyProperties(dto, contributor);
        return contributor;
    }
}

File: src/main/kotlin/com/indiepost/mapper/ContributorMapper.kt
package com.indiepost.mapper

import com.indiepost.dto.ContributorDto
import com.indiepost.enums.Types
import com.indiepost.model.Contributor
import java.time.LocalDateTime

fun Contributor.createDto(): ContributorDto {
    val dto = ContributorDto()
    dto.id = id
    dto.fullName = fullName
    dto.email = email
    dto.isEmailVisible = isEmailVisible
    dto.subEmail = subEmail
    dto.phone = phone
    dto.isPhoneVisible = isPhoneVisible
    dto.picture = picture
    dto.isPictureVisible = isPictureVisible
    dto.url = url
    dto.title = title
    dto.isTitleVisible = isTitleVisible
    dto.description = description
    dto.isDescriptionVisible = isDescriptionVisible
    contributorType?.let {
        dto.contributorType = it.toString()
    }
    displayType?.let {
        dto.displayType = it.toString()
    }
    dto.lastUpdated = lastUpdated
    dto.created = created
    dto.etc = etc
    return dto
}

fun Contributor.merge(dto: ContributorDto) {
    fullName = dto.fullName
    title = dto.title
    isTitleVisible = dto.isTitleVisible
    picture = dto.picture
    isPictureVisible = dto.isPictureVisible
    dto.email?.let {
        email = it
    }
    subEmail = dto.subEmail
    isEmailVisible = dto.isEmailVisible
    description = dto.description
    isDescriptionVisible = dto.isDescriptionVisible
    dto.contributorType?.let {
        contributorType = Types.ContributorType.valueOf(it)
    }
    dto.displayType?.let {
        displayType = Types.ContributorDisplayType.valueOf(it)
    }
    etc = dto.etc
    phone = dto.phone
    isPhoneVisible = dto.isPhoneVisible
    url = dto.url
    isUrlVisible = dto.isUrlVisible
    lastUpdated = LocalDateTime.now()
}

fun ContributorDto.createEntity(): Contributor {
    val contributor = Contributor()
    contributor.created = LocalDateTime.now()
    contributor.merge(this)
    return contributor
}

